{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Procedural Notes  \n",
    "\n",
    "### Enzyme/Pathway Selection  \n",
    "Kegg glycolysis/Glucogenesis-Reference Pathway:  \n",
    "\n",
    ">Enzymes selected for each pathway are:  \n",
    ">- Glycolysis: K00844, K11645, K01689, K16370  \n",
    ">- Citric Acid Cycle: K01647, K00030, K01900, K00116  \n",
    ">- Pentose Phosphate Pathway: K13937, K00036, K01807, K06859  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating the genes, pathways and enzymes table with respective assigned columns\n",
    "import sqlite3\n",
    "conn = sqlite3.connect('my.db')\n",
    "c = conn.cursor()\n",
    "# to avoid conflict with previously built table\n",
    "c.execute(\"\"\"DROP TABLE genes\"\"\")\n",
    "c.execute(\"\"\"DROP TABLE pathways\"\"\")\n",
    "c.execute(\"\"\"DROP TABLE enzymes\"\"\")\n",
    "\n",
    "# Caution: use INTEGER instead of INT for primary key generation\n",
    "c.execute(\"\"\"CREATE TABLE genes (id INTEGER PRIMARY KEY, name TEXT, description TEXT, organism TEXT, nucleotide_sequence TEXT,\n",
    "chromosome TEXT, location TEXT, start INT, end INT, strand VARCHAR(1), translated_sequence TEXT);\"\"\")\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE pathways (id INTEGER PRIMARY KEY, name TEXT, description TEXT);\"\"\")\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE enzymes (id INTEGER PRIMARY KEY, kegg_id TEXT, name TEXT, description TEXT, function TEXT, ec_number TEXT);\"\"\")\n",
    "\n",
    "conn.commit()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inserting values into the pathway table\n",
    "1. get the names of the pathways\n",
    "2. search up the description of each pathway\n",
    "3. Take in two list with equal length and insert the into the corresponding field"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "conn = sqlite3.connect('my.db')\n",
    "c = conn.cursor()\n",
    "cmd = ''' INSERT INTO pathways (name, description) VALUES(?,?)'''\n",
    "\n",
    "names = [\"Glycolysis\",\"Citric Acid Cycle\",\"Pentose Phosphate\"]\n",
    "descriptions = [\"the breakdown of glucose by enzymes, releasing energy and pyruvic acid.\" , \"the sequence of reactions by which most living cells generate energy during the process of aerobic respiration. It takes place in the mitochondria, consuming oxygen, producing carbon dioxide and water as waste products, and converting ADP to energy-rich ATP\",\"a metabolic pathway parallel to glycolysis, It generates NADPH and pentoses (5-carbon sugars) as well as ribose 5-phosphate, the last one a precursor for the synthesis of nucleotides\"]\n",
    "\n",
    "\n",
    "# Take in two list with equal length and insert the into the corresponding field\n",
    "for i in range(len(names)):\n",
    "    name = names[i]\n",
    "    description = descriptions[i]\n",
    "    temp_value = (name,description)\n",
    "    c.execute(cmd,temp_value)\n",
    "    conn.commit()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query the name and ec_number of the selected Enzymes from KEGG\n",
    "1. Make a sample.list file in the current directory with the following lines  \n",
    "K00844   \n",
    "K11645   \n",
    "K01689  \n",
    "K16370  \n",
    "K01647  \n",
    "K00030  \n",
    "K01900  \n",
    "K00116  \n",
    "K13937  \n",
    "K00036  \n",
    "K01807  \n",
    "K06859  \n",
    "\n",
    "2. In the terminal execute: `curl -g -s -S http://rest.kegg.jp/list/ko | grep -f sample.list | sed \"s/cpd\\://\" > sample_extracted.table.txt`\n",
    "\n",
    "3. Inspect the output file sample_extracted.table.txt  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ko:K00030\tIDH3; isocitrate dehydrogenase (NAD+) [EC:1.1.1.41]\n",
      "\n",
      "ko:K00036\tG6PD, zwf; glucose-6-phosphate 1-dehydrogenase [EC:1.1.1.49 1.1.1.363]\n",
      "\n",
      "ko:K00116\tmqo; malate dehydrogenase (quinone) [EC:1.1.5.4]\n",
      "\n",
      "ko:K00844\tHK; hexokinase [EC:2.7.1.1]\n",
      "\n",
      "ko:K01647\tCS, gltA; citrate synthase [EC:2.3.3.1]\n",
      "\n",
      "ko:K01689\tENO, eno; enolase [EC:4.2.1.11]\n",
      "\n",
      "ko:K01807\trpiA; ribose 5-phosphate isomerase A [EC:5.3.1.6]\n",
      "\n",
      "ko:K01900\tLSC2; succinyl-CoA synthetase beta subunit [EC:6.2.1.4 6.2.1.5]\n",
      "\n",
      "ko:K06859\tpgi1; glucose-6-phosphate isomerase, archaeal [EC:5.3.1.9]\n",
      "\n",
      "ko:K11645\tfbaB; fructose-bisphosphate aldolase, class I [EC:4.1.2.13]\n",
      "\n",
      "ko:K13937\tH6PD; hexose-6-phosphate dehydrogenase [EC:1.1.1.47 3.1.1.31]\n",
      "\n",
      "ko:K16370\tpfkB; 6-phosphofructokinase 2 [EC:2.7.1.11]\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(['ko:K00030',\n",
       "  'ko:K00036',\n",
       "  'ko:K00116',\n",
       "  'ko:K00844',\n",
       "  'ko:K01647',\n",
       "  'ko:K01689',\n",
       "  'ko:K01807',\n",
       "  'ko:K01900',\n",
       "  'ko:K06859',\n",
       "  'ko:K11645',\n",
       "  'ko:K13937',\n",
       "  'ko:K16370'],\n",
       " ['IDH3',\n",
       "  'G6PD,zwf',\n",
       "  'mqo',\n",
       "  'HK',\n",
       "  'CS,gltA',\n",
       "  'ENO,eno',\n",
       "  'rpiA',\n",
       "  'LSC2',\n",
       "  'pgi1',\n",
       "  'fbaB',\n",
       "  'H6PD',\n",
       "  'pfkB'],\n",
       " ['isocitrate dehydrogenase (NAD+)',\n",
       "  'glucose-6-phosphate 1-dehydrogenase',\n",
       "  'malate dehydrogenase (quinone)',\n",
       "  'hexokinase',\n",
       "  'citrate synthase',\n",
       "  'enolase',\n",
       "  'ribose 5-phosphate isomerase A',\n",
       "  'succinyl-CoA synthetase beta subunit',\n",
       "  'glucose-6-phosphate isomerase, archaeal',\n",
       "  'fructose-bisphosphate aldolase, class I',\n",
       "  'hexose-6-phosphate dehydrogenase',\n",
       "  '6-phosphofructokinase 2'],\n",
       " ['EC:1.1.1.41]',\n",
       "  'EC:1.1.1.49 1.1.1.363]',\n",
       "  'EC:1.1.5.4]',\n",
       "  'EC:2.7.1.1]',\n",
       "  'EC:2.3.3.1]',\n",
       "  'EC:4.2.1.11]',\n",
       "  'EC:5.3.1.6]',\n",
       "  'EC:6.2.1.4 6.2.1.5]',\n",
       "  'EC:5.3.1.9]',\n",
       "  'EC:4.1.2.13]',\n",
       "  'EC:1.1.1.47 3.1.1.31]',\n",
       "  'EC:2.7.1.11]'])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# parse in the sample_extracted.table.txt for enzymes table insertion\n",
    "def parse_extracted_table():\n",
    "    filename = \"sample_extracted.table.txt\"\n",
    "    in_file = open(filename, \"rt\")\n",
    "    lines = []\n",
    "    # separate each line in the for for inspection and downstream analysis\n",
    "    for line in in_file:\n",
    "        lines.append(line.rstrip('\\n'))\n",
    "        print(line)\n",
    "    return lines\n",
    "\n",
    "lines = parse_extracted_table()\n",
    "\n",
    "\n",
    "# the get balues function split each line in the sample_extracted.table.txt to render the kegg_id, \n",
    "# enzyme name, function, and ec_number as separate lists\n",
    "def get_values(lines):\n",
    "    kegg_ids = []\n",
    "    names = []\n",
    "    functions = []\n",
    "    ec_numbers = []\n",
    "    for line in lines:\n",
    "        # first, separate out the ec_number by the \"[\" symbol\n",
    "        line_ec = line.split(\"[\")\n",
    "        ec_number = line_ec[-1]\n",
    "        # separate the by the \";\" symbol, the first half will be kegg_id and the name, the second part function\n",
    "        line_2_part = line_ec[0].split(\";\")\n",
    "        # line_part1 contains kegg_id and the name\n",
    "        line_part1 = line_2_part[0].split()\n",
    "        # line_part2 contains function description\n",
    "        line_part2 = line_2_part[1].split()\n",
    "#         print(line_part1)\n",
    "#         print(line_part2)\n",
    "        kegg_id = line_part1[0]\n",
    "        # join the 1:end of part1 to form the name\n",
    "        temp_name = line_part1[1:]\n",
    "        temp_name[:] = [''.join(temp_name[:])]\n",
    "        name = temp_name[0]\n",
    "        # join all segments in part2 to form the function string\n",
    "        temp_function = line_part2[:]\n",
    "        temp_function[:] = [\" \".join(temp_function[:])]\n",
    "        function = temp_function[0]     \n",
    "        kegg_ids.append(kegg_id)\n",
    "        names.append(name)\n",
    "        functions.append(function)\n",
    "        ec_numbers.append(ec_number)\n",
    "    return kegg_ids,names,functions,ec_numbers\n",
    "\n",
    "\n",
    "# inspect the 3 output lists\n",
    "get_values(lines)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert values into the enzymes table with 4 equal length list names, functions, and kegg_ids and ec_numbers \n",
    "# into the corresponding columns of the enzymes table\n",
    "import sqlite3\n",
    "conn = sqlite3.connect('my.db')\n",
    "c = conn.cursor()\n",
    "cmd = ''' INSERT INTO enzymes (kegg_id, name, description, function, ec_number) VALUES(?,?,?,?,?)'''\n",
    "\n",
    "# get the four lists\n",
    "kegg_ids,names,functions,ec_numbers = get_values(lines)\n",
    "descriptions = ['isocitrate dehydrogenase', 'glucose-6-phosphate 1-dehydrogenase', 'malate dehydrogenase', 'hexokinase', 'citrate synthase', 'enolase', 'ribose 5-phosphate isomerase', 'succinyl-CoA synthetase', 'glucose-6-phosphate isomerase', 'fructose-bisphosphate aldolase', 'hexose-6-phosphate dehydrogenase', '6-phosphofructokinase']\n",
    "functions = [\"Glycolysis\"]*4+[\"Citric Acid Cycle\"]*4 + [\"Pentose Phosphate\"]*4\n",
    "# insertion\n",
    "for i in range(len(names)):\n",
    "    name = names[i]\n",
    "    function = functions[i]\n",
    "    description = descriptions[i]\n",
    "    kegg_id = kegg_ids[i]\n",
    "    ec_number = ec_numbers[i]\n",
    "    temp_value = (kegg_id, name, description, function, ec_number)\n",
    "    c.execute(cmd,temp_value)\n",
    "    conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "term_list\n",
      "['Drosophila[ORGN] isocitrate dehydrogenase', 'Drosophila[ORGN] glucose-6-phosphate 1-dehydrogenase', 'Drosophila[ORGN] malate dehydrogenase', 'Drosophila[ORGN] hexokinase', 'Drosophila[ORGN] citrate synthase', 'Drosophila[ORGN] enolase', 'Drosophila[ORGN] ribose 5-phosphate isomerase', 'Drosophila[ORGN] succinyl-CoA synthetase', 'Drosophila[ORGN] glucose-6-phosphate isomerase', 'Drosophila[ORGN] fructose-bisphosphate aldolase', 'Drosophila[ORGN] hexose-6-phosphate dehydrogenase', 'Drosophila[ORGN] 6-phosphofructokinase', 'Escherichia coli[ORGN] isocitrate dehydrogenase', 'Escherichia coli[ORGN] glucose-6-phosphate 1-dehydrogenase', 'Escherichia coli[ORGN] malate dehydrogenase', 'Escherichia coli[ORGN] hexokinase', 'Escherichia coli[ORGN] citrate synthase', 'Escherichia coli[ORGN] enolase', 'Escherichia coli[ORGN] ribose 5-phosphate isomerase', 'Escherichia coli[ORGN] succinyl-CoA synthetase', 'Escherichia coli[ORGN] glucose-6-phosphate isomerase', 'Escherichia coli[ORGN] fructose-bisphosphate aldolase', 'Escherichia coli[ORGN] hexose-6-phosphate dehydrogenase', 'Escherichia coli[ORGN] 6-phosphofructokinase', 'Homo sapiens[ORGN] isocitrate dehydrogenase', 'Homo sapiens[ORGN] glucose-6-phosphate 1-dehydrogenase', 'Homo sapiens[ORGN] malate dehydrogenase', 'Homo sapiens[ORGN] hexokinase', 'Homo sapiens[ORGN] citrate synthase', 'Homo sapiens[ORGN] enolase', 'Homo sapiens[ORGN] ribose 5-phosphate isomerase', 'Homo sapiens[ORGN] succinyl-CoA synthetase', 'Homo sapiens[ORGN] glucose-6-phosphate isomerase', 'Homo sapiens[ORGN] fructose-bisphosphate aldolase', 'Homo sapiens[ORGN] hexose-6-phosphate dehydrogenase', 'Homo sapiens[ORGN] 6-phosphofructokinase']\n",
      "name_list\n",
      "['IDH3', 'G6PD,zwf', 'mqo', 'HK', 'CS,gltA', 'ENO,eno', 'rpiA', 'LSC2', 'pgi1', 'fbaB', 'H6PD', 'pfkB', 'IDH3', 'G6PD,zwf', 'mqo', 'HK', 'CS,gltA', 'ENO,eno', 'rpiA', 'LSC2', 'pgi1', 'fbaB', 'H6PD', 'pfkB', 'IDH3', 'G6PD,zwf', 'mqo', 'HK', 'CS,gltA', 'ENO,eno', 'rpiA', 'LSC2', 'pgi1', 'fbaB', 'H6PD', 'pfkB']\n",
      "organism_list\n",
      "['Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens']\n",
      "descriptions_list\n",
      "['isocitrate dehydrogenase', 'glucose-6-phosphate 1-dehydrogenase', 'malate dehydrogenase', 'hexokinase', 'citrate synthase', 'enolase', 'ribose 5-phosphate isomerase', 'succinyl-CoA synthetase', 'glucose-6-phosphate isomerase', 'fructose-bisphosphate aldolase', 'hexose-6-phosphate dehydrogenase', '6-phosphofructokinase', 'isocitrate dehydrogenase', 'glucose-6-phosphate 1-dehydrogenase', 'malate dehydrogenase', 'hexokinase', 'citrate synthase', 'enolase', 'ribose 5-phosphate isomerase', 'succinyl-CoA synthetase', 'glucose-6-phosphate isomerase', 'fructose-bisphosphate aldolase', 'hexose-6-phosphate dehydrogenase', '6-phosphofructokinase', 'isocitrate dehydrogenase', 'glucose-6-phosphate 1-dehydrogenase', 'malate dehydrogenase', 'hexokinase', 'citrate synthase', 'enolase', 'ribose 5-phosphate isomerase', 'succinyl-CoA synthetase', 'glucose-6-phosphate isomerase', 'fructose-bisphosphate aldolase', 'hexose-6-phosphate dehydrogenase', '6-phosphofructokinase']\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "#Entrez.esearch(db,term, sort)\n",
    "#Entrez.efetch(db, id, rettype, retmode)\n",
    "#for enzymes and genes only; Entrez doesn't have pathway info\n",
    "\n",
    "\n",
    "# a helper function to search and fetch for gene, nucleotide, and protein by specifying the db_type and the term\n",
    "def search_fetch(db_type,term):\n",
    "    Entrez.email = 'ych323@berkeley.edu'\n",
    "    handle = Entrez.esearch(db = db_type,\n",
    "                       term = term,\n",
    "                       sort= 'relevance',\n",
    "                       idtype= 'acc')\n",
    "    fetched_dict = Entrez.read(handle)\n",
    "#     print(fetched_dict)\n",
    "    if not fetched_dict[\"IdList\"]:\n",
    "        data = \"None\"\n",
    "        return data\n",
    "    handle = Entrez.efetch(db = db_type, id = fetched_dict[\"IdList\"][0], rettype = 'fasta', retmode = 'text')\n",
    "    data = handle.read()\n",
    "    return data\n",
    "\n",
    "\n",
    "# build a term list of all the search \"terms species[ORGN] enzyme_name\"\n",
    "# build the 3 * 12 entries of name, organism, and description of the gene\n",
    "species_list = [\"Drosophila\",\"Escherichia coli\",\"Homo sapiens\"]\n",
    "ter_list = [str(species) + \"[ORGN]\" + \" \" for species in species_list ]\n",
    "term_list = [str(term) + str(description) for term in ter_list for description in descriptions]\n",
    "name_list = [str(name) for term in ter_list for name in names]\n",
    "organism_list = [str(species) for species in species_list for name in names]\n",
    "descriptions_list = [str(description) for species in species_list for description in descriptions]\n",
    "\n",
    "# for the strand column, the DNA nucleotide sequence stored in GeneBank are all 5' to 3' of the sense strand, and so the \n",
    "# sign of the strand should all be \"+\"\n",
    "strand_list = ['+']*len(term_list)\n",
    "\n",
    "# inspect the built lists \n",
    "print(\"term_list\")\n",
    "print(term_list)\n",
    "print(\"name_list\")\n",
    "print(name_list)\n",
    "print(\"organism_list\")\n",
    "print(organism_list)\n",
    "print(\"descriptions_list\")\n",
    "print(descriptions_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Drosophila[ORGN] isocitrate dehydrogenase\n",
      "parsed\n",
      "Drosophila[ORGN] glucose-6-phosphate 1-dehydrogenase\n",
      "parsed\n",
      "Drosophila[ORGN] malate dehydrogenase\n",
      "parsed\n",
      "Drosophila[ORGN] hexokinase\n",
      "parsed\n",
      "Drosophila[ORGN] citrate synthase\n",
      "parsed\n",
      "Drosophila[ORGN] enolase\n",
      "parsed\n",
      "Drosophila[ORGN] ribose 5-phosphate isomerase\n",
      "parsed\n",
      "Drosophila[ORGN] succinyl-CoA synthetase\n",
      "parsed\n",
      "Drosophila[ORGN] glucose-6-phosphate isomerase\n",
      "parsed\n",
      "Drosophila[ORGN] fructose-bisphosphate aldolase\n",
      "parsed\n",
      "Drosophila[ORGN] hexose-6-phosphate dehydrogenase\n",
      "Drosophila[ORGN] 6-phosphofructokinase\n",
      "parsed\n",
      "Escherichia coli[ORGN] isocitrate dehydrogenase\n",
      "parsed\n",
      "Escherichia coli[ORGN] glucose-6-phosphate 1-dehydrogenase\n",
      "parsed\n",
      "Escherichia coli[ORGN] malate dehydrogenase\n",
      "parsed\n",
      "Escherichia coli[ORGN] hexokinase\n",
      "parsed\n",
      "Escherichia coli[ORGN] citrate synthase\n",
      "parsed\n",
      "Escherichia coli[ORGN] enolase\n",
      "parsed\n",
      "Escherichia coli[ORGN] ribose 5-phosphate isomerase\n",
      "parsed\n",
      "Escherichia coli[ORGN] succinyl-CoA synthetase\n",
      "parsed\n",
      "Escherichia coli[ORGN] glucose-6-phosphate isomerase\n",
      "parsed\n",
      "Escherichia coli[ORGN] fructose-bisphosphate aldolase\n",
      "parsed\n",
      "Escherichia coli[ORGN] hexose-6-phosphate dehydrogenase\n",
      "Escherichia coli[ORGN] 6-phosphofructokinase\n",
      "parsed\n",
      "Homo sapiens[ORGN] isocitrate dehydrogenase\n",
      "parsed\n",
      "Homo sapiens[ORGN] glucose-6-phosphate 1-dehydrogenase\n",
      "parsed\n",
      "Homo sapiens[ORGN] malate dehydrogenase\n",
      "parsed\n",
      "Homo sapiens[ORGN] hexokinase\n",
      "parsed\n",
      "Homo sapiens[ORGN] citrate synthase\n",
      "parsed\n",
      "Homo sapiens[ORGN] enolase\n",
      "parsed\n",
      "Homo sapiens[ORGN] ribose 5-phosphate isomerase\n",
      "parsed\n",
      "Homo sapiens[ORGN] succinyl-CoA synthetase\n",
      "parsed\n",
      "Homo sapiens[ORGN] glucose-6-phosphate isomerase\n",
      "parsed\n",
      "Homo sapiens[ORGN] fructose-bisphosphate aldolase\n",
      "parsed\n",
      "Homo sapiens[ORGN] hexose-6-phosphate dehydrogenase\n",
      "parsed\n",
      "Homo sapiens[ORGN] 6-phosphofructokinase\n",
      "parsed\n",
      "peak nucleotide_list\n",
      "['CCAAAATTTCTTTGTTTCTATTGGGCAGTCAAAATTTTAAATTTTATTAAAATTAATATTGATAATGCTACCCATCGATCCCAACAATGAAGTCGCCTATTCGTTTGTTGTGTTTGGCGCATCGGGTCGTTTGGCGCTGAGCAGAATCTTTCCGGCGCTCTGGCAACTCTATCGCGATAATCGTCTGCCCCAGGCTACCAAAATCTTTACATTCTGCCGCACCAAGCTCGAGCTTAAAACGTTTCGCATCAAGTGCGTGCCGTACATGGGCCTGGAGAGGGACCGCGATCCAATCTTGTACAATAGCTTCTGGTCCAATGTGCATTGCTGTGACGGCCAATACGACAACCCGGAGCACTATGCGAATCTTACGGCGGCCATGGTGCGCCAGGAGAAGAGGCACGATATGCGGTTTGCCAATCGCATATTCTATCTCGCTCTGCCGCCAGTCGTCTTCGACCAGGTGACCCTCAATATCGCGCGCAAGTGTCTCTCAGCCACCGGCTGGAATCACGTCGTCCTGGAGAAGCCCTTTGCCAGAAATGATATAACCTACAAAGTATATCAGACACAGCTCTGCCAGACGTTCAAGGAGTCCCAGATCTATATGATCGACCACTATCTGAGCAGGCAAGTGATTCACAATATGTTTGTGCTGCGCTTCACCAATCTCATCTGGGCGGACACGTGGAACAACAAGCACATCGCAGCGATTATGATTAGTGTGAAGACCGAGCTCCCAGTGCAGGCGCGCGCTGAATATTTCAATCAGTTTGGCATCATACGCGACGTTATGACCAACCACATGATGCAGCTGGTGGCAATGCTCGCCATCGAGCAGCCGTACAGCAGCAACGTGGAGGACATGCGCGACGAGAAGCTGAAGGTGCTTAAGCAAGTGCTGAACGTCGATATGAGCGACGTGTTGCTGGGGCAGTACTGCAACAACGGCAGGGAGTCTGATCCCGTCAAGGTGGGCTACACGCAGCACTCCTACATACCCAAGGACTCGCTGACGCCCACCTTTGCGATGGTTGTGCTGCACATCAATAGCAAGCGCTGGCGCAGCGTGCCCTTCATCTTGCGCGTGGGCAAAGCGCTCAACGAAACCAAGACGGAGGTGCGAGTGCAGTACAAATCGACGGACTGCAATCACAGTGGCAGCAGCTCGAGTATACCAAATGAGCTGGTGTTGCGTCTGTCGCCGCGCGAACAGCTCCTCATGCGCATGATGCAGAAGCGACCGGGGCCGCACATGTGTCTCAGGGAAACGGAACTGGACCTGATCCTGCGCGATAGAGCATTGCCCATCAACTATCAGGCTCTGCTACTCGACATCTTTGCTGGCAATCAGATGATGTTCATGCGCACCGATGAGCAGTGCGAAATCTGGCGCATATTCTCGCCAGTGCTGGCCGCAATCGAGTATGAGCGCACGAAACCCATACTCTATGACTTTGGCTCTTGCGGACCCATGGCCGCATACGAAATGGCAGCCAGACACGGCTTTAAATTTTACCAATCTGACGAATGGCATGCCAGCCGGGGCAACCCCAATGCCGATCCGACCGTCCCTGAGAATTCCAAGAAACCAGCTGTCAATCGAGCCCTGCTAGCGACACCAGCTTGGATGAATCCAGTCACGTAGGGTAGCGGAAGAAGCTAAGCTTTTGTCTAAAGTCAGTTTTAGTTATAGTCCCTTCTGGTATATATGTATATATATAGCAGATGCATTTATGTCATAAATTGGTTCAACGCAAATTTAATAAAATTTACCACGCAAAAGTTTTTTAATTAACCTCCT', 'AAAAAGTAGTGACTGGAAAATAATTGTGAAAAGTAACAGGTTCCCAATTGGCGATCCCTATTTACGTTTTTACTTGAAGTTCGTCCGTGGAGCGTTGCACCTGTGAAATTCTAAAATGGCTGAACCAATTCGTGTTGTGGTGACCGGAGCCGCTGGCCAAATCGCCTACTCCCTGCTGTACATGATTGCCCGCGGCGAGGTGTTCGGCAAGGATCAGCCCATCGTGCTGCACCTGCTCGACATTCCGCCCATGGTTGGTGTGCTCGAGGGCGTGGTCATGGAATTGGCCGACTGCGCTCTGCCGCTGCTGGTCGAGGTGGTGCCCACCACCGACCCGGCTGTTGGATTCAAGGATGTCTCGGCCGCTTTCCTCGTGGGCGCCATGCCCCGCAAGGAGGGAATGGAGCGAAAGGATCTGCTGTCCGCCAACGTGAAGATCTTCAGGACCCAGGGCCAGGCTCTGGACAAGTTCGCCAAGAAGGACGTCAAGGTGCTGGTTGTGGGCAACCCGGCCAATACCAATGCCCTGGTCTGCTCCTCCTATGCGCCTTCCATTCCGCGCGAAAATTTCTCGGCTATGACTCGCCTGGATCAGAATCGCGCCACATCCCAGATCGCCGCCAAATTGGGTGTGCCAATTTCTGCTGTTAAGAACATCATTATTTGGGGCAACCACTCCTCCACCCAGTATCCCGATGCTGGACAGGCCAAGGTGACCGCCAATGGTACCGTTAAGTCCGTGGTGGATGCGATCAACGATAATGGCTACCTGCAGGGATCCTTCGTGGAAACCGTTCAGAAGCGTGGAGCCGCTGTCATTGCGGCCCGCAAAATGTCATCGGCCATGTCGGCGGCGAAGGCCGCTTGCGATCACATGCACGACTGGTGGAACGGCACTGCTCCCGGCCAGTTCGTCTCCATGGGTGTCTTCTCCGACGGCAGCTACGATTCGCCCAAGGATGTGATCTTCTCGTTCCCCGTCGAGATCAAGAATAAGCAATGGAAAATCGTCTCTGGCCTGACCCTCAGCGATTTCGCTAAGACAAAGTTAAGCGTTACCGGCAAGGAGCTGCAGGAGGAGAAGGACGAGGCTCTGTCCGTGCTGGACTCGAATGTGTCCAACTTGTAATGCGATCCTTAAACGTTAATTACTAACATCATAGTTGTGCATTTAGGAACAAAGTTCATGGCAATAAAAGTTTTGTAAATAGCTCTGAAATGGTTTTGTTTTGTGCAAGTTAATTGTTGCTAGCGATTTGAACTATAAGATTAAAGGCTTAATGGCTGAT', 'ATGCCATTTGTGGACCCCTCAGCGTCGCACATATACACGCCATATCTCCAACCATGCCGCCCCAAAACCGATTTTCAGTTTTTGACCTTCAAGGTTCCTTCACGGCGAGCGAGAGAGAGAGAGAGTGAGAGTGTGAGCGCGGCTCGCAGTCGCGGTCGGCGCCGGCAGCGCAGCAGCAGCGACAGCGAAAAGTGAGCACCAGTCGGCGAGTGAATTTAATTTCGTAAAAAACGTTCCGAGCGTTTGGACGCGTGCAGCGGACGAAAATAAAACACGGCGAGAGCGAGAATCGTGTGTGTGGACACTCAAAGACTCAAAATTGAAACCAACTCGAAAAGGTTGCACTGCAAATGCCCAAGTAGAAAAGCCAACGCGAGTGTGCACATCGAATACGAAAATTGTAAACGTTTTTTTTTTGCCGTGTGAGTGCAAGCGCATAAGATCGAAGAAAAACAGCCAGAGCAAGAAAGAACCGAAAAAAGGCAGAGAGAATCAAAAAAGAAAAAAGGAAAAAAAAAACAAGCAGGAAAACACCAACATAATAATGGATCGGGAGCTGAACGGGGCGATGAAGAGCGTGTCCATAAATAGCGCCCAACTGAATGGTCACAATAACGGTGGCGGGCTAGATGAAACGGACCGCGTGGGCGTGGCACAGGGGGCGGCACTGAGTGCAACAGAGAAAATTACAACAACGACGGCGGCAGCTACAAAATCAGCGACAGCGACCACAAATGCAACCACAGCAACAGCCACAACCACAAATCTTACAACTCACAGCCCGCAGCAAATTGCACTCTTATCCGCTGCAGAGAAAAGTAAGATGGTGCACGAGTTATGTCAGCAGCTGCTTCTAACGGACGAACAGGTGCAGGAGCTGTGCTACCGCATCCTGCACGAGCTGCGCCGCGGCCTGGCCAAGGATACACATCCCAAGGCGAATGTCAAGTGCTTCGTAACGTATGTGCAGGATCTGCCCAATGGCAATGAGCGTGGCAAGTTCTTGGCCTTGGATCTGGGTGGCACCAACTTCCGGGTGCTGCTCATCCATCTGCAGGAGAACAACGATTTTCAAATGGAGTCCCGCATCTATGCCATACCGCAGCACATTATGATTGGATCGGGTACACAGCTATTCGATCACATCGCCGAATGTCTGTCCAACTTTATGGCCGAGCACAATGTGTACAAGGAGCGTTTGCCACTCGGTTTCACCTTCTCGTTCCCGCTGCGCCAATTGGGTCTGACCAAGGGCCTGCTGGAGACCTGGACCAAGGGCTTCAACTGCGCTGGCGTCGTCAACGAGGATGTTGTCCAGTTGCTAAAGGATGCCATCGCACGGCGCGGCGATGTCCAGATCGATGTGTGTGCCATACTCAATGACACCACTGGCACGCTGATGAGCTGCGCCTGGAAGAATCACAACTGCAAGATCGGTCTGATTGTCGGCACCGGTGCAAATGCCTGCTATATGGAGCGTGTGGAGGAGGCGGAGCTCTTCGCCGCCGAGGATCCGCGCAAGAAGCACGTCCTGATCAACACGGAATGGGGCGCATTCGGCGACAATGGCGCACTGGACTTTGTGCGCACCGAATTCGATCGGGACATCGATGTGCACAGCATCAATCCGGGCAAGCAGACGTTCGAGAAGATGATATCGGGCATGTATATGGGCGAGCTGGTGCGTCTGGTGCTGGTGAAGATGACCCAGGCGGGTATACTCTTCAATGGACAGGATTCCGAAGTGCTAAACACCCGCGGACTGTTCTTCACCAAGTATGTGAGCGAAATTGAAGCTGACGAGCCAGGCAACTTCACCAATTGCCGCCTGGTGCTGGAGGAGCTCGGTCTGACCAATGCCACCGACGGGGACTGCGCCAATGTGCGGTACATCTGCGAGTGCGTCTCGAAACGGGCTGCCCACCTGGTGTCCGCCGGCATTGCCACGCTGATCAACAAAATGGACGAGCCAACGGTGACGGTGGGCGTCGATGGCAGTGTGTACCGCTTCCATCCCAAGTTCCATAACCTGATGGTCGAGAAGATCTCGCAGCTGATCAAGCCGGGCATCACCTTCGATCTGATGCTGTCCGAGGACGGTTCCGGTCGTGGAGCTGCCCTGGTGGCGGCCGTCGCCTGTCGCGAAGACATACTCAATGGCAAGAAGTAGGATGGATAGGGAGCTGGAGCTGGATCAGGAACCAGGAGGAGGAGCCGCTGGAGCAGCTGGAGCAGTCCTAAGAGATACAGCAACACACCCACACAACCACAAAAACAGCAGAGCAAATGAGAAATACCTAGTAAACGGCCACGGTGGCGTATGCGTAATAATTACAGTAGAGCCACGCTAACTCACAAAAGCCCTATAGCCAAACGATATGAAAAAATGAAGCTGCCACGACGAAAAGTAATCGAAAAGCCCACGAAAAACTTATATGTGCATACGTAGGGCATTTGTGAGTTGGCGCGACTTGGCTTTAGTATTTTAATTAGGCAAAAACCAGGTTTTTCACTACAAACACAAGCGTAACAGAAACCCCCCCACTAAGAGACAGAAGATCTGTGGCATTCTATCTCTTTCTCTCCTAGTAAAAGAAAACAAAGAAGAAGAAAAAACACTATGCTTAATGATTATGGATCGATTTTCGATTTTTGTTGAAAATTTGTTATGCTATTATATATATATATATTTTACAAGAAAACATAATATACACACACTAAAAACATAATGTATATGTTTGCGTATATATAGTACAATTTGTGAAGCGAATTTCTTTTCGATTTCGATTGAAACAACAAATTTTTTTTGGTAATTCCCTGCCCAAAAGACCCCTGAAACCCCCCTAAAAAAACGGAGATACTTAAGAGCAAGTGAGAAAAGAAACTGTTATACTACTGTGCTACAAGCATTCTAGCATATCCTGCAGGACAACGGAAAGAGAAACTGTAAAACCGCACATTGTTAAAGCAAAAACGAAAACGAAAAAAGCAACAAAAAACAAGCCGCTAGAGAAGAGACAGAAACAGAGGGATAGAAAGGAAGAGAGCGAGAGAGAGCGAGCGAGAGAAAGAGAGAGAGGATTACTAGGTAATGCTACGGTGTCCTTTCTGTAATTGTAAATGTTTATATAGAACCTATTACGAAGCGTTATATGATAATACTGCATACATATATATAAATATAAATATGATATATTTAGCCACTATATTACTGATTTTTTTTTTTGTAATGTATTGCTTGTCTTTGTTTAGTGAATTTGTTTGTTGTTAACACTTAAGCAGGTAAACTCTAAACAAAAATAAACTGTTTTTTTTTTTCCTTTTGATATAGAAACGCACTTTGTGAGCTATTATTTATGTAATTGTACATTTGACCTAGTTTAGTGTTTAGTTCTTATCTTAGTTTCTAAATGTCGTCGAAGAAGCCAGCAATACAACCAATATAATACCCCCACCCCAAAAAAAAAACCCAACCCACGCAGCCCTTAATTATATTTAATTATTGTTTACGAACAACAAACGAAGAGAATGCGAGAAGCTAAAAGGCAAATAAGCAAAATGGAAAGCAAATAAAATTTGTTCTTAATTCATACACAAGTAGTTGTTACTTATTTCTCCCAAATTAATTGGTTTTCCAAAGTGATTAGGTTTTTTCCAGATTGATTTCAGTATGATTTGATCTGTTTTCCATATGATTCGGTATCATTTGATCATTTTTGCAGCTGCTGGCCGGCATGCTCAGTCAACACAAAAACTTAAACAATTCGCAATTGAGGCAAAACGAAGATACATTTTTTTGGTCTAATACTAAGCGGAGAATGAGCCTAACCTAAGAGATATGTTATAATAAATGGCGGGAGCAGCGAACCAGTGTTTCTATTTTTCAAGAAAGAAAAAAACGAAAAACAAGAAGAATGTAATATAAACTTAGCCCATGTTAAAATGAAAATTGAAAACCAATTAAAAACTCTCGTTATCAAAAGGCGATTCAAACATAAATTGCTAAGAACAAACAAACACAAATCGAAGGCTTGAAAGTGCAGATGAGATGAATTTAGTAAATGATATAAACAAATGGAATGTTTGTCAACAGTGAAGCAAAATAAAGACGATTGAACCAAACAACAGAGAAGCAAATC']\n"
     ]
    }
   ],
   "source": [
    "# Get the nucleotide sequences of each enzyme through searching entrez and fetch the most relevent result\n",
    "def get_nucleotide_sequence(term_list):\n",
    "    nucleotide_list = []\n",
    "    for term in term_list:\n",
    "        print(term)\n",
    "        data = search_fetch(\"nucleotide\",term)\n",
    "        # If data is \"None\", then the entrez search has fetched no id and the idlist is empty, and \n",
    "        # so cannot be parsed by SeqIO, so should be directly append to the nucleotide list\n",
    "        if data == \"None\":\n",
    "            nucleotide_list.append(data)\n",
    "            continue\n",
    "            \n",
    "        # If data is not \"None\", then the search_and_fetch function return the most relevent sequence, wrtie data into a\n",
    "        # fasta file and then use SeqIO.parse to get the sequence to a string for sqlite insertion\n",
    "        f = open('temp.fa','w')\n",
    "        f.write(data)\n",
    "        f.close()\n",
    "        print(\"parsed\")\n",
    "        for seq in SeqIO.parse(\"temp.fa\", \"fasta\"):\n",
    "            nucleotide_list.append(str(seq.seq))\n",
    "    return nucleotide_list\n",
    "\n",
    "nucleotide_list = get_nucleotide_sequence(term_list)\n",
    "# inspect first three entries of the nucleotide_list\n",
    "\n",
    "print(\"peak nucleotide_list\")\n",
    "print(nucleotide_list[1:4])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Drosophila[ORGN] isocitrate dehydrogenase\n",
      "parsed\n",
      "Drosophila[ORGN] glucose-6-phosphate 1-dehydrogenase\n",
      "parsed\n",
      "Drosophila[ORGN] malate dehydrogenase\n",
      "parsed\n",
      "Drosophila[ORGN] hexokinase\n",
      "parsed\n",
      "Drosophila[ORGN] citrate synthase\n",
      "parsed\n",
      "Drosophila[ORGN] enolase\n",
      "parsed\n",
      "Drosophila[ORGN] ribose 5-phosphate isomerase\n",
      "parsed\n",
      "Drosophila[ORGN] succinyl-CoA synthetase\n",
      "parsed\n",
      "Drosophila[ORGN] glucose-6-phosphate isomerase\n",
      "parsed\n",
      "Drosophila[ORGN] fructose-bisphosphate aldolase\n",
      "parsed\n",
      "Drosophila[ORGN] hexose-6-phosphate dehydrogenase\n",
      "Drosophila[ORGN] 6-phosphofructokinase\n",
      "parsed\n",
      "Escherichia coli[ORGN] isocitrate dehydrogenase\n",
      "parsed\n",
      "Escherichia coli[ORGN] glucose-6-phosphate 1-dehydrogenase\n",
      "parsed\n",
      "Escherichia coli[ORGN] malate dehydrogenase\n",
      "parsed\n",
      "Escherichia coli[ORGN] hexokinase\n",
      "parsed\n",
      "Escherichia coli[ORGN] citrate synthase\n",
      "parsed\n",
      "Escherichia coli[ORGN] enolase\n",
      "parsed\n",
      "Escherichia coli[ORGN] ribose 5-phosphate isomerase\n",
      "parsed\n",
      "Escherichia coli[ORGN] succinyl-CoA synthetase\n",
      "parsed\n",
      "Escherichia coli[ORGN] glucose-6-phosphate isomerase\n",
      "parsed\n",
      "Escherichia coli[ORGN] fructose-bisphosphate aldolase\n",
      "parsed\n",
      "Escherichia coli[ORGN] hexose-6-phosphate dehydrogenase\n",
      "Escherichia coli[ORGN] 6-phosphofructokinase\n",
      "parsed\n",
      "Homo sapiens[ORGN] isocitrate dehydrogenase\n",
      "parsed\n",
      "Homo sapiens[ORGN] glucose-6-phosphate 1-dehydrogenase\n",
      "parsed\n",
      "Homo sapiens[ORGN] malate dehydrogenase\n",
      "parsed\n",
      "Homo sapiens[ORGN] hexokinase\n",
      "parsed\n",
      "Homo sapiens[ORGN] citrate synthase\n",
      "parsed\n",
      "Homo sapiens[ORGN] enolase\n",
      "parsed\n",
      "Homo sapiens[ORGN] ribose 5-phosphate isomerase\n",
      "parsed\n",
      "Homo sapiens[ORGN] succinyl-CoA synthetase\n",
      "parsed\n",
      "Homo sapiens[ORGN] glucose-6-phosphate isomerase\n",
      "parsed\n",
      "Homo sapiens[ORGN] fructose-bisphosphate aldolase\n",
      "parsed\n",
      "Homo sapiens[ORGN] hexose-6-phosphate dehydrogenase\n",
      "parsed\n",
      "Homo sapiens[ORGN] 6-phosphofructokinase\n",
      "parsed\n",
      "peek protein_list\n",
      "['MLPIDPNNEVAYSFVVFGASGRLALSRIFPALWQLYRDNRLPQATKIFTFCRTKLELKTFRIKCVPYMGLERDRDPILYNSFWSNVHCCDGQYDNPEHYANLTAAMVRQEKRHDMRFANRIFYLALPPVVFDQVTLNIARKCLSATGWNHVVLEKPFARNDITYKVYQTQLCQTFKESQIYMIDHYLSRQVIHNMFVLRFTNLIWADTWNNKHIAAIMISVKTELPVQARAEYFNQFGIIRDVMTNHMMQLVAMLAIEQPYSSNVEDMRDEKLKVLKQVLNVDMSDVLLGQYCNNGRESDPVKVGYTQHSYIPKDSLTPTFAMVVLHINSKRWRSVPFILRVGKALNETKTEVRVQYKSTDCNHSGSSSSIPNELVLRLSPREQLLMRMMQKRPGPHMCLRETELDLILRDRALPINYQALLLDIFAGNQMMFMRTDEQCEIWRIFSPVLAAIEYERTKPILYDFGSCGPMAAYEMAARHGFKFYQSDEWHASRGNPNADPTVPENSKKPAVNRALLATPAWMNPVT', 'MAEPIRVVVTGAAGQIAYSLLYMIARGEVFGKDQPIVLHLLDIPPMVGVLEGVVMELADCALPLLVEVVPTTDPAVGFKDVSAAFLVGAMPRKEGMERKDLLSANVKIFRTQGQALDKFAKKDVKVLVVGNPANTNALVCSSYAPSIPRENFSAMTRLDQNRATSQIAAKLGVPISAVKNIIIWGNHSSTQYPDAGQAKVTANGTVKSVVDAINDNGYLQGSFVETVQKRGAAVIAARKMSSAMSAAKAACDHMHDWWNGTAPGQFVSMGVFSDGSYDSPKDVIFSFPVEIKNKQWKIVSGLTLSDFAKTKLSVTGKELQEEKDEALSVLDSNVSNL']\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "\n",
    "# following similar approach as get_nucleotide to get the protein sequence from entrez search and fetch\n",
    "def get_protein(term_list):\n",
    "    protein_list = []\n",
    "    for term in term_list:\n",
    "        print(term)\n",
    "        data = search_fetch(\"protein\",term)\n",
    "        if data == \"None\":\n",
    "            protein_list.append(data)\n",
    "            continue\n",
    "        f = open('temp.fa','w')\n",
    "        f.write(data)\n",
    "        f.close()\n",
    "        print(\"parsed\")\n",
    "        for seq in SeqIO.parse(\"temp.fa\", \"fasta\"):\n",
    "            protein_list.append(str(seq.seq))\n",
    "    return protein_list\n",
    "protein_list = get_protein(term_list)\n",
    "\n",
    "#inspect the output protein list\n",
    "print(\"peek protein_list\")\n",
    "print(protein_list[1:3])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Chromosome: 3L; Location: 3-24 cM', 'Chromosome: Un', 'Chromosome: X', 'Chromosome: X; Location: 1-28 cM', 'Chromosome: X', 'Chromosome: 4; Location: 4- cM', 'Chromosome: Un', 'Chromosome: 3L; Location: 3-9 cM', 'Chromosome: Un', 'Chromosome: Un', 'None', 'Chromosome: Un', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'None', 'not_in_entrez', 'Chromosome: 2; Location: 2q34', 'Chromosome: 1; Location: 1p36.22', 'Chromosome: 7; Location: 7q11.23', 'Chromosome: 10; Location: 10q22.1', 'Chromosome: 12; Location: 12q13.3', 'Chromosome: 12; Location: 12p13.31', 'Chromosome: 2; Location: 2p11.2', 'Chromosome: 13; Location: 13q14.2', 'Chromosome: 19; Location: 19q13.11', 'Chromosome: 16; Location: 16p11.2', 'Chromosome: 1; Location: 1p36.22', 'Chromosome: 12; Location: 12q13.11']\n",
      "['Annotation: Chromosome 3L NT_037436.4 (8356404..8360906, complement)', 'not_in_entrez', 'not_in_entrez', 'Annotation: Chromosome X NC_004354.4 (9585675..9589813)', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'Annotation: Chromosome 3L NT_037436.4 (3814211..3818383)', 'not_in_entrez', 'not_in_entrez', 'None', 'not_in_entrez', 'Annotation:  NC_000913.3 (1195123..1196373)', 'Annotation:  NC_000913.3 (1934839..1936314, complement)', 'Annotation:  NC_000913.3 (3383330..3384268, complement)', 'Annotation:  NC_000913.3 (410144..411052)', 'Annotation:  NC_000913.3 (753185..754468, complement)', 'Annotation:  NC_025179.1 (57883..58233)', 'Annotation:  NC_000913.3 (4313350..4313799)', 'Annotation:  NC_000913.3 (764180..765049)', 'Annotation:  NC_017634.1 (4289297..4290946)', 'Annotation:  NC_000913.3 (3070165..3071244, complement)', 'None', 'Annotation:  NC_000913.3 (4107552..4108514)', 'Annotation: Chromosome 2 NC_000002.12 (208236227..208255143, complement)', 'Annotation: Chromosome 1 NC_000001.11 (9234767..9271337)', 'Annotation: Chromosome 7 NC_000007.14 (76048019..76067508)', 'Annotation: Chromosome 10 NC_000010.11 (69269991..69401882)', 'Annotation: Chromosome 12 NC_000012.12 (56271699..56300391, complement)', 'Annotation: Chromosome 12 NC_000012.12 (6914450..6923696)', 'Annotation: Chromosome 2 NC_000002.12 (88691658..88750935)', 'Annotation: Chromosome 13 NC_000013.11 (47942656..48001326, complement)', 'Annotation: Chromosome 19 NC_000019.10 (34353330..34402413)', 'Annotation: Chromosome 16 NC_000016.10 (30053090..30070420)', 'Annotation: Chromosome 1 NC_000001.11 (9234767..9271337)', 'Annotation: Chromosome 12 NC_000012.12 (48105253..48146404)']\n"
     ]
    }
   ],
   "source": [
    "# to get the chromosome, start, and end position and possibly other information from the gene entrez search\n",
    "# CAUTION: hasn't implemented through iteration yet\n",
    "from Bio import Entrez\n",
    "# function to separate lines of the gene search and fetch output txt file\n",
    "def parse_gene_entrez():\n",
    "    filename = \"temp_gene_entrez.txt\"\n",
    "    in_file = open(filename, \"rt\")\n",
    "    lines = []\n",
    "    # separate each line in the for for inspection and downstream analysis\n",
    "    for line in in_file:\n",
    "        lines.append(line.rstrip('\\n'))\n",
    "    return lines\n",
    "\n",
    "def find_loci(lines):\n",
    "    locus = \"not_in_entrez\"\n",
    "    for line in lines:\n",
    "        if \"Chromosome:\" in line:\n",
    "            locus = line\n",
    "            break\n",
    "    return locus\n",
    "\n",
    "def find_annotation(lines):\n",
    "    annotation = \"not_in_entrez\"\n",
    "    for line in lines:\n",
    "        if \"Annotation:\" in line:\n",
    "            annotation = line\n",
    "            break\n",
    "    return annotation\n",
    "\n",
    "def get_location(term_list):   \n",
    "    loci = []\n",
    "    annotations = []\n",
    "    Entrez.email = 'ych323@berkeley.edu'\n",
    "    for term in term_list:\n",
    "        data = search_fetch(\"gene\",term)\n",
    "        if data == \"None\":\n",
    "            locus = \"None\"\n",
    "            annotation = \"None\"\n",
    "        else:\n",
    "            f = open('temp_gene_entrez.txt','w')\n",
    "            f.write(data)\n",
    "            f.close()\n",
    "            lines = parse_gene_entrez()\n",
    "            #print(lines)\n",
    "            # Location and annotation are in the 5 and 6 of the lines list \n",
    "            # if there's no desired field, then append \"insufficient info\"\n",
    "            locus = find_loci(lines)\n",
    "            annotation = find_annotation(lines)\n",
    "        loci.append(locus)\n",
    "        annotations.append(annotation)\n",
    "    return loci, annotations\n",
    "\n",
    "loci, annotations = get_location(term_list)\n",
    "\n",
    "print(loci)\n",
    "print(annotations)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosomes:\n",
      "[' 3L', 'unknown', ' X', ' X', ' X', ' 4', 'unknown', ' 3L', 'unknown', 'unknown', 'None', 'unknown', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', ' 2', ' 1', ' 7', ' 10', ' 12', ' 12', ' 2', ' 13', ' 19', ' 16', ' 1', ' 12']\n",
      "locations:\n",
      "[' 3-24 cM', 'unknown', 'not in entrez', ' 1-28 cM', 'not in entrez', ' 4- cM', 'unknown', ' 3-9 cM', 'unknown', 'unknown', 'None', 'unknown', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', 'circular prokaryotic DNA', ' 2q34', ' 1p36.22', ' 7q11.23', ' 10q22.1', ' 12q13.3', ' 12p13.31', ' 2p11.2', ' 13q14.2', ' 19q13.11', ' 16p11.2', ' 1p36.22', ' 12q13.11']\n"
     ]
    }
   ],
   "source": [
    "# parse the chromosome and location from loci list and the start and end position from the annotations site\n",
    "def parse_chromosome(loci):\n",
    "    chromosomes = []\n",
    "    locations = []\n",
    "    for temp_string in loci: \n",
    "        if \"Chromosome:\" not in temp_string:\n",
    "            location = temp_string\n",
    "            chromosome = temp_string\n",
    "        elif \"Un\" in temp_string:\n",
    "            location = \"unknown\"\n",
    "            chromosome = \"unknown\"\n",
    "        else:\n",
    "            #print(temp_string)\n",
    "            temp_list = temp_string.split(\";\")\n",
    "            chromosome = temp_list[0].split(\":\")[1]\n",
    "            try:\n",
    "                location = temp_list[1].split(\":\")[1]\n",
    "            except:\n",
    "                location = 'not in entrez'\n",
    "        locations.append(location)\n",
    "        chromosomes.append(chromosome)\n",
    "    return chromosomes, locations\n",
    "\n",
    "chromosomes, locations =  parse_chromosome(loci)\n",
    "\n",
    "# to account that E.coli has no chromosome:\n",
    "chromosomes[12:24] = ['circular prokaryotic DNA']*12\n",
    "locations[12:24] = ['circular prokaryotic DNA']*12\n",
    "print(\"chromosomes:\")\n",
    "print(chromosomes)\n",
    "print(\"locations:\")\n",
    "print(locations)\n",
    "# from the list we can see that at lot of genes does not have info in all organisms\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "starts:\n",
      "['8356404', 'not_in_entrez', 'not_in_entrez', '9585675', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', '3814211', 'not_in_entrez', 'not_in_entrez', 'None', 'not_in_entrez', '1195123', '1934839', '3383330', '410144', '753185', '57883', '4313350', '764180', '4289297', '3070165', 'None', '4107552', '208236227', '9234767', '76048019', '69269991', '56271699', '6914450', '88691658', '47942656', '34353330', '30053090', '9234767', '48105253']\n",
      "ends:\n",
      "['8360906', 'not_in_entrez', 'not_in_entrez', '9589813', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', '3818383', 'not_in_entrez', 'not_in_entrez', 'None', 'not_in_entrez', '1196373', '1936314', '3384268', '411052', '754468', '58233', '4313799', '765049', '4290946', '3071244', 'None', '4108514', '208255143', '9271337', '76067508', '69401882', '56300391', '6923696', '88750935', '48001326', '34402413', '30070420', '9271337', '48146404']\n"
     ]
    }
   ],
   "source": [
    "def parse_annotation(annotations):\n",
    "    starts = []\n",
    "    ends = []\n",
    "    for temp_string in annotations: \n",
    "        if \"Annotation:\" not in temp_string:\n",
    "            start = temp_string\n",
    "            end = temp_string\n",
    "        elif \"Un\" in temp_string:\n",
    "            start = temp_string\n",
    "            end = temp_string\n",
    "        else:\n",
    "            #print(temp_string)\n",
    "            a = temp_string\n",
    "            b = a.split(\"(\")\n",
    "            c = b[-1].split(\")\")\n",
    "            d = c[0].split(\" \")\n",
    "            start_end = d[0].split(\"..\")\n",
    "            start = start_end[0]\n",
    "            end = start_end[1].split(\",\")[0]\n",
    "        starts.append(start)\n",
    "        ends.append(end)\n",
    "    return starts, ends\n",
    "\n",
    "starts, ends = parse_annotation(annotations)\n",
    "print(\"starts:\")\n",
    "print(starts)\n",
    "print(\"ends:\")\n",
    "print(ends)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sanity check on the entrez gene output\n",
    "> as we have all the lists for the gene table insertion, check that the length of the lists are consistent\n",
    "\n",
    "- id INTEGER PRIMARY KEY  \n",
    "- name TEXT  \n",
    "- description TEXT  \n",
    "- organism TEXT  \n",
    "- nucleotide_sequence TEXT  \n",
    "- chromosome TEXT  \n",
    "- location TEXT  \n",
    "- start INT  \n",
    "- end INT  \n",
    "- strand VARCHAR(1)  \n",
    "- translated_sequence TEXT  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36\n",
      "36\n",
      "36\n",
      "36\n",
      "36\n",
      "36\n",
      "36\n",
      "36\n",
      "36\n",
      "36\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "\n",
    "print(len(name_list))\n",
    "print(len(organism_list))\n",
    "print(len(descriptions_list))\n",
    "print(len(strand_list))\n",
    "print(len(nucleotide_list))\n",
    "print(len(protein_list))\n",
    "print(len(chromosomes))\n",
    "print(len(locations))\n",
    "print(len(starts))\n",
    "print(len(ends))\n",
    "# all the list are in same order with the same length\n",
    "\n",
    "# manually fix the None fields in cds and aa sequence the H6PD gene\n",
    "h6pd_nt = str([s.seq for s in SeqIO.parse(\"H6PD_nt.fa\", \"fasta\")][0])\n",
    "h6pd_aa = str([s.seq for s in SeqIO.parse(\"H6PD_aa.fa\", \"fasta\")][0])   \n",
    "for i in range(len(name_list)):\n",
    "    if nucleotide_list[i] == 'None':\n",
    "        nucleotide_list[i] = h6pd_nt\n",
    "        protein_list[i] = h6pd_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "conn = sqlite3.connect('my.db')\n",
    "c = conn.cursor()\n",
    "cmd = ''' INSERT INTO genes (name,description,organism, nucleotide_sequence,chromosome,location,start,end,strand,translated_sequence) VALUES(?,?,?,?,?,?,?,?,?,?)'''\n",
    "\n",
    "# insertion\n",
    "for i in range(len(name_list)):\n",
    "    name = name_list[i]\n",
    "    description = descriptions_list[i]\n",
    "    organism = organism_list[i]\n",
    "    nucleotide_sequence = nucleotide_list[i]\n",
    "    chromosome = chromosomes[i]\n",
    "    location = locations[i]\n",
    "    start = starts[i]\n",
    "    end = ends[i]\n",
    "    strand = strand_list[i]\n",
    "    translated_sequence = protein_list[i]\n",
    "    temp_value = (name,description,organism, nucleotide_sequence,chromosome,location,start,end,strand,translated_sequence)\n",
    "    c.execute(cmd,temp_value)\n",
    "    conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Demo of the functionality of three individual tables\n",
    "  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inspect the pathways table\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Glycolysis</td>\n",
       "      <td>the breakdown of glucose by enzymes, releasing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Citric Acid Cycle</td>\n",
       "      <td>the sequence of reactions by which most living...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Pentose Phosphate</td>\n",
       "      <td>a metabolic pathway parallel to glycolysis, It...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id               name                                        description\n",
       "0   1         Glycolysis  the breakdown of glucose by enzymes, releasing...\n",
       "1   2  Citric Acid Cycle  the sequence of reactions by which most living...\n",
       "2   3  Pentose Phosphate  a metabolic pathway parallel to glycolysis, It..."
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import sqlite3\n",
    "conn = sqlite3.connect(\"my.db\")\n",
    "\n",
    "print(\"inspect the pathways table\")\n",
    "df1 = pd.read_sql_query(\"select * from pathways;\", conn)\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inspect the enzymes table\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>function</th>\n",
       "      <th>ec_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ko:K00030</td>\n",
       "      <td>IDH3</td>\n",
       "      <td>isocitrate dehydrogenase</td>\n",
       "      <td>Glycolysis</td>\n",
       "      <td>EC:1.1.1.41]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>ko:K00036</td>\n",
       "      <td>G6PD,zwf</td>\n",
       "      <td>glucose-6-phosphate 1-dehydrogenase</td>\n",
       "      <td>Glycolysis</td>\n",
       "      <td>EC:1.1.1.49 1.1.1.363]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>ko:K00116</td>\n",
       "      <td>mqo</td>\n",
       "      <td>malate dehydrogenase</td>\n",
       "      <td>Glycolysis</td>\n",
       "      <td>EC:1.1.5.4]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>ko:K00844</td>\n",
       "      <td>HK</td>\n",
       "      <td>hexokinase</td>\n",
       "      <td>Glycolysis</td>\n",
       "      <td>EC:2.7.1.1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>ko:K01647</td>\n",
       "      <td>CS,gltA</td>\n",
       "      <td>citrate synthase</td>\n",
       "      <td>Citric Acid Cycle</td>\n",
       "      <td>EC:2.3.3.1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>ko:K01689</td>\n",
       "      <td>ENO,eno</td>\n",
       "      <td>enolase</td>\n",
       "      <td>Citric Acid Cycle</td>\n",
       "      <td>EC:4.2.1.11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>ko:K01807</td>\n",
       "      <td>rpiA</td>\n",
       "      <td>ribose 5-phosphate isomerase</td>\n",
       "      <td>Citric Acid Cycle</td>\n",
       "      <td>EC:5.3.1.6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>ko:K01900</td>\n",
       "      <td>LSC2</td>\n",
       "      <td>succinyl-CoA synthetase</td>\n",
       "      <td>Citric Acid Cycle</td>\n",
       "      <td>EC:6.2.1.4 6.2.1.5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>ko:K06859</td>\n",
       "      <td>pgi1</td>\n",
       "      <td>glucose-6-phosphate isomerase</td>\n",
       "      <td>Pentose Phosphate</td>\n",
       "      <td>EC:5.3.1.9]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>ko:K11645</td>\n",
       "      <td>fbaB</td>\n",
       "      <td>fructose-bisphosphate aldolase</td>\n",
       "      <td>Pentose Phosphate</td>\n",
       "      <td>EC:4.1.2.13]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>ko:K13937</td>\n",
       "      <td>H6PD</td>\n",
       "      <td>hexose-6-phosphate dehydrogenase</td>\n",
       "      <td>Pentose Phosphate</td>\n",
       "      <td>EC:1.1.1.47 3.1.1.31]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>ko:K16370</td>\n",
       "      <td>pfkB</td>\n",
       "      <td>6-phosphofructokinase</td>\n",
       "      <td>Pentose Phosphate</td>\n",
       "      <td>EC:2.7.1.11]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id    kegg_id      name                          description  \\\n",
       "0    1  ko:K00030      IDH3             isocitrate dehydrogenase   \n",
       "1    2  ko:K00036  G6PD,zwf  glucose-6-phosphate 1-dehydrogenase   \n",
       "2    3  ko:K00116       mqo                 malate dehydrogenase   \n",
       "3    4  ko:K00844        HK                           hexokinase   \n",
       "4    5  ko:K01647   CS,gltA                     citrate synthase   \n",
       "5    6  ko:K01689   ENO,eno                              enolase   \n",
       "6    7  ko:K01807      rpiA         ribose 5-phosphate isomerase   \n",
       "7    8  ko:K01900      LSC2              succinyl-CoA synthetase   \n",
       "8    9  ko:K06859      pgi1        glucose-6-phosphate isomerase   \n",
       "9   10  ko:K11645      fbaB       fructose-bisphosphate aldolase   \n",
       "10  11  ko:K13937      H6PD     hexose-6-phosphate dehydrogenase   \n",
       "11  12  ko:K16370      pfkB                6-phosphofructokinase   \n",
       "\n",
       "             function               ec_number  \n",
       "0          Glycolysis            EC:1.1.1.41]  \n",
       "1          Glycolysis  EC:1.1.1.49 1.1.1.363]  \n",
       "2          Glycolysis             EC:1.1.5.4]  \n",
       "3          Glycolysis             EC:2.7.1.1]  \n",
       "4   Citric Acid Cycle             EC:2.3.3.1]  \n",
       "5   Citric Acid Cycle            EC:4.2.1.11]  \n",
       "6   Citric Acid Cycle             EC:5.3.1.6]  \n",
       "7   Citric Acid Cycle     EC:6.2.1.4 6.2.1.5]  \n",
       "8   Pentose Phosphate             EC:5.3.1.9]  \n",
       "9   Pentose Phosphate            EC:4.1.2.13]  \n",
       "10  Pentose Phosphate   EC:1.1.1.47 3.1.1.31]  \n",
       "11  Pentose Phosphate            EC:2.7.1.11]  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"inspect the enzymes table\")\n",
    "df2 = pd.read_sql_query(\"select * from enzymes;\", conn)\n",
    "df2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Design of Associative tables\n",
    "1. enzymes_in_pathways: a 2 columns table that link the enzyme id to pathway id  \n",
    "2. genes_in_enzyme: a 2 columns table that link the gene id to enzyme id\n",
    "2. genes_in_pathway: a 2 columns table that link the gene id to pathway id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inspect the first four genes in table with human \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IDH3</td>\n",
       "      <td>isocitrate dehydrogenase</td>\n",
       "      <td>2</td>\n",
       "      <td>208236227</td>\n",
       "      <td>208255143</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G6PD,zwf</td>\n",
       "      <td>glucose-6-phosphate 1-dehydrogenase</td>\n",
       "      <td>1</td>\n",
       "      <td>9234767</td>\n",
       "      <td>9271337</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mqo</td>\n",
       "      <td>malate dehydrogenase</td>\n",
       "      <td>7</td>\n",
       "      <td>76048019</td>\n",
       "      <td>76067508</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HK</td>\n",
       "      <td>hexokinase</td>\n",
       "      <td>10</td>\n",
       "      <td>69269991</td>\n",
       "      <td>69401882</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       name                          description chromosome      start  \\\n",
       "0      IDH3             isocitrate dehydrogenase          2  208236227   \n",
       "1  G6PD,zwf  glucose-6-phosphate 1-dehydrogenase          1    9234767   \n",
       "2       mqo                 malate dehydrogenase          7   76048019   \n",
       "3        HK                           hexokinase         10   69269991   \n",
       "\n",
       "         end strand  \n",
       "0  208255143      +  \n",
       "1    9271337      +  \n",
       "2   76067508      +  \n",
       "3   69401882      +  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"inspect the first four genes in table with human \")\n",
    "df3 = pd.read_sql_query(\"select name,description,chromosome,start,end,strand from genes where organism = 'Homo sapiens' limit 4\", conn)\n",
    "df3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inspect the first four genes in genes table with E.coli \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IDH3</td>\n",
       "      <td>isocitrate dehydrogenase</td>\n",
       "      <td>circular prokaryotic DNA</td>\n",
       "      <td>1195123</td>\n",
       "      <td>1196373</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G6PD,zwf</td>\n",
       "      <td>glucose-6-phosphate 1-dehydrogenase</td>\n",
       "      <td>circular prokaryotic DNA</td>\n",
       "      <td>1934839</td>\n",
       "      <td>1936314</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mqo</td>\n",
       "      <td>malate dehydrogenase</td>\n",
       "      <td>circular prokaryotic DNA</td>\n",
       "      <td>3383330</td>\n",
       "      <td>3384268</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HK</td>\n",
       "      <td>hexokinase</td>\n",
       "      <td>circular prokaryotic DNA</td>\n",
       "      <td>410144</td>\n",
       "      <td>411052</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       name                          description                chromosome  \\\n",
       "0      IDH3             isocitrate dehydrogenase  circular prokaryotic DNA   \n",
       "1  G6PD,zwf  glucose-6-phosphate 1-dehydrogenase  circular prokaryotic DNA   \n",
       "2       mqo                 malate dehydrogenase  circular prokaryotic DNA   \n",
       "3        HK                           hexokinase  circular prokaryotic DNA   \n",
       "\n",
       "     start      end strand  \n",
       "0  1195123  1196373      +  \n",
       "1  1934839  1936314      +  \n",
       "2  3383330  3384268      +  \n",
       "3   410144   411052      +  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"inspect the first four genes in genes table with E.coli \")\n",
    "df4 = pd.read_sql_query(\"select name,description,chromosome,start,end,strand from genes where organism = 'Escherichia coli' limit 4\", conn)\n",
    "df4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inspect the first three genes in genes table with Drosophila \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IDH3</td>\n",
       "      <td>isocitrate dehydrogenase</td>\n",
       "      <td>3L</td>\n",
       "      <td>8356404</td>\n",
       "      <td>8360906</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HK</td>\n",
       "      <td>hexokinase</td>\n",
       "      <td>X</td>\n",
       "      <td>9585675</td>\n",
       "      <td>9589813</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LSC2</td>\n",
       "      <td>succinyl-CoA synthetase</td>\n",
       "      <td>3L</td>\n",
       "      <td>3814211</td>\n",
       "      <td>3818383</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   name               description chromosome    start      end strand\n",
       "0  IDH3  isocitrate dehydrogenase         3L  8356404  8360906      +\n",
       "1    HK                hexokinase          X  9585675  9589813      +\n",
       "2  LSC2   succinyl-CoA synthetase         3L  3814211  3818383      +"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"inspect the first three genes in genes table with Drosophila \")\n",
    "df5 = pd.read_sql_query(\"select name,description,chromosome,start,end,strand from genes where organism = 'Drosophila' and chromosome != 'None' and start != 'not_in_entrez' limit 5\", conn)\n",
    "df5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inspect the nucleotide sequence\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>IDH3</td>\n",
       "      <td>ACGCCGATTGCGCAACAATTTGACGATTGTACGCCACCCAAGAGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>G6PD,zwf</td>\n",
       "      <td>CCAAAATTTCTTTGTTTCTATTGGGCAGTCAAAATTTTAAATTTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>mqo</td>\n",
       "      <td>AAAAAGTAGTGACTGGAAAATAATTGTGAAAAGTAACAGGTTCCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>HK</td>\n",
       "      <td>ATGCCATTTGTGGACCCCTCAGCGTCGCACATATACACGCCATATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>CS,gltA</td>\n",
       "      <td>GAATTCGTCAGAAATGAGCTAAACAAATTTAAATCATTAAATGCGA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id      name                                nucleotide_sequence\n",
       "0   1      IDH3  ACGCCGATTGCGCAACAATTTGACGATTGTACGCCACCCAAGAGAG...\n",
       "1   2  G6PD,zwf  CCAAAATTTCTTTGTTTCTATTGGGCAGTCAAAATTTTAAATTTTA...\n",
       "2   3       mqo  AAAAAGTAGTGACTGGAAAATAATTGTGAAAAGTAACAGGTTCCCA...\n",
       "3   4        HK  ATGCCATTTGTGGACCCCTCAGCGTCGCACATATACACGCCATATC...\n",
       "4   5   CS,gltA  GAATTCGTCAGAAATGAGCTAAACAAATTTAAATCATTAAATGCGA..."
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"inspect the nucleotide sequence\")\n",
    "df6 = pd.read_sql_query(\"select id, name, nucleotide_sequence from genes limit 5\", conn)\n",
    "df6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inspect the translated sequence\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>translated_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>IDH3</td>\n",
       "      <td>MFALRRTAAMMTSVHRQPHLSQAVFRANFAISAADPKTHKISMDNE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>G6PD,zwf</td>\n",
       "      <td>MLPIDPNNEVAYSFVVFGASGRLALSRIFPALWQLYRDNRLPQATK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>mqo</td>\n",
       "      <td>MAEPIRVVVTGAAGQIAYSLLYMIARGEVFGKDQPIVLHLLDIPPM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>HK</td>\n",
       "      <td>MVHELCQQLLLTDEQVQELCYRILHELRRGLAKDTHPKANVKCFVT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>CS,gltA</td>\n",
       "      <td>MSLYRISARKLSEAQKLPNVGAYVRMIAADGKSLRDVLAAKVPQEQ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id      name                                translated_sequence\n",
       "0   1      IDH3  MFALRRTAAMMTSVHRQPHLSQAVFRANFAISAADPKTHKISMDNE...\n",
       "1   2  G6PD,zwf  MLPIDPNNEVAYSFVVFGASGRLALSRIFPALWQLYRDNRLPQATK...\n",
       "2   3       mqo  MAEPIRVVVTGAAGQIAYSLLYMIARGEVFGKDQPIVLHLLDIPPM...\n",
       "3   4        HK  MVHELCQQLLLTDEQVQELCYRILHELRRGLAKDTHPKANVKCFVT...\n",
       "4   5   CS,gltA  MSLYRISARKLSEAQKLPNVGAYVRMIAADGKSLRDVLAAKVPQEQ..."
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"inspect the translated sequence\")\n",
    "df7 = pd.read_sql_query(\"select id, name, translated_sequence from genes limit 5\", conn)\n",
    "df7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "conn = sqlite3.connect('my.db')\n",
    "c = conn.cursor()\n",
    "# to avoid conflict with previously built table\n",
    "c.execute(\"\"\"DROP TABLE enzymes_in_pathways\"\"\")\n",
    "c.execute(\"\"\"DROP TABLE genes_in_enzymes\"\"\")\n",
    "c.execute(\"\"\"DROP TABLE genes_in_pathways\"\"\")\n",
    "# Caution: use INTEGER instead of INT for primary key generation\n",
    "c.execute(\"\"\"CREATE TABLE enzymes_in_pathways (pathway_id INT, enzyme_id INT);\"\"\")\n",
    "c.execute(\"\"\"CREATE TABLE genes_in_enzymes (enzyme_id INT, gene_id INT);\"\"\")\n",
    "c.execute(\"\"\"CREATE TABLE genes_in_pathways (pathway_id INT, gene_id INT);\"\"\")\n",
    "conn.commit()\n",
    "\n",
    "cmd = ''' INSERT INTO enzymes_in_pathways (pathway_id,enzyme_id) VALUES(?,?)'''\n",
    "c = conn.cursor()\n",
    "# build enzymes_in_pathways\n",
    "pathways = [1]*4 + [2]*4 + [3]*4 \n",
    "enzymes = list(range(1,13))\n",
    "for i in range(len(pathways)):\n",
    "    temp_val = (pathways[i],enzymes[i])\n",
    "    c.execute(cmd,temp_val)\n",
    "    conn.commit()\n",
    "\n",
    "# build genes_in_enzymes\n",
    "cmd = ''' INSERT INTO genes_in_enzymes (enzyme_id,gene_id) VALUES(?,?)'''\n",
    "enzymes = list(range(1,13))*3\n",
    "genes = list(range(1,37))\n",
    "for i in range(len(genes)):\n",
    "    temp_val = (enzymes[i],genes[i])\n",
    "    c.execute(cmd,temp_val)\n",
    "    conn.commit()\n",
    "\n",
    "# build genes_in_enzymes\n",
    "cmd = ''' INSERT INTO genes_in_pathways (pathway_id,gene_id) VALUES(?,?)'''\n",
    "pathways = [1,1,1,1,2,2,2,2,3,3,3,3]*3\n",
    "genes = list(range(1,37))\n",
    "for i in range(len(genes)):\n",
    "    temp_val = (pathways[i],genes[i])\n",
    "    c.execute(cmd,temp_val)\n",
    "    conn.commit()\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Demo of the use of associative tables\n",
    "- through associative tables, we can select a group of more specific items by grouping them with a more general item"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "select enzyme by a specific pathway: all enzyme in Glycolysis pathway\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>function</th>\n",
       "      <th>ec_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ko:K00030</td>\n",
       "      <td>IDH3</td>\n",
       "      <td>isocitrate dehydrogenase</td>\n",
       "      <td>Glycolysis</td>\n",
       "      <td>EC:1.1.1.41]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>ko:K00036</td>\n",
       "      <td>G6PD,zwf</td>\n",
       "      <td>glucose-6-phosphate 1-dehydrogenase</td>\n",
       "      <td>Glycolysis</td>\n",
       "      <td>EC:1.1.1.49 1.1.1.363]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>ko:K00116</td>\n",
       "      <td>mqo</td>\n",
       "      <td>malate dehydrogenase</td>\n",
       "      <td>Glycolysis</td>\n",
       "      <td>EC:1.1.5.4]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>ko:K00844</td>\n",
       "      <td>HK</td>\n",
       "      <td>hexokinase</td>\n",
       "      <td>Glycolysis</td>\n",
       "      <td>EC:2.7.1.1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id    kegg_id      name                          description    function  \\\n",
       "0   1  ko:K00030      IDH3             isocitrate dehydrogenase  Glycolysis   \n",
       "1   2  ko:K00036  G6PD,zwf  glucose-6-phosphate 1-dehydrogenase  Glycolysis   \n",
       "2   3  ko:K00116       mqo                 malate dehydrogenase  Glycolysis   \n",
       "3   4  ko:K00844        HK                           hexokinase  Glycolysis   \n",
       "\n",
       "                ec_number  \n",
       "0            EC:1.1.1.41]  \n",
       "1  EC:1.1.1.49 1.1.1.363]  \n",
       "2             EC:1.1.5.4]  \n",
       "3             EC:2.7.1.1]  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"select enzyme by a specific pathway: all enzyme in Glycolysis pathway\")\n",
    "df8 = pd.read_sql_query(\"select * from enzymes where id in (select enzyme_id from enzymes_in_pathways where pathway_id = 1)\", conn)\n",
    "df8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "select genes by a specific pathway: all human enzymes in Glycolysis pathway\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IDH3</td>\n",
       "      <td>isocitrate dehydrogenase</td>\n",
       "      <td>2</td>\n",
       "      <td>208236227</td>\n",
       "      <td>208255143</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G6PD,zwf</td>\n",
       "      <td>glucose-6-phosphate 1-dehydrogenase</td>\n",
       "      <td>1</td>\n",
       "      <td>9234767</td>\n",
       "      <td>9271337</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mqo</td>\n",
       "      <td>malate dehydrogenase</td>\n",
       "      <td>7</td>\n",
       "      <td>76048019</td>\n",
       "      <td>76067508</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HK</td>\n",
       "      <td>hexokinase</td>\n",
       "      <td>10</td>\n",
       "      <td>69269991</td>\n",
       "      <td>69401882</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       name                          description chromosome      start  \\\n",
       "0      IDH3             isocitrate dehydrogenase          2  208236227   \n",
       "1  G6PD,zwf  glucose-6-phosphate 1-dehydrogenase          1    9234767   \n",
       "2       mqo                 malate dehydrogenase          7   76048019   \n",
       "3        HK                           hexokinase         10   69269991   \n",
       "\n",
       "         end strand  \n",
       "0  208255143      +  \n",
       "1    9271337      +  \n",
       "2   76067508      +  \n",
       "3   69401882      +  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"select genes by a specific pathway: all human enzymes in Glycolysis pathway\")\n",
    "df9 = pd.read_sql_query(\"select name,description,chromosome,start,end,strand from genes where organism = 'Homo sapiens' and id in (select gene_id from genes_in_pathways where pathway_id = 1)\", conn)\n",
    "df9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "select genes by a specific enzyme: genes from different species for the HK enzyme\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>organism</th>\n",
       "      <th>nucleotide_sequence</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>location</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>translated_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>HK</td>\n",
       "      <td>hexokinase</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>ATGCCATTTGTGGACCCCTCAGCGTCGCACATATACACGCCATATC...</td>\n",
       "      <td>X</td>\n",
       "      <td>1-28 cM</td>\n",
       "      <td>9585675</td>\n",
       "      <td>9589813</td>\n",
       "      <td>+</td>\n",
       "      <td>MVHELCQQLLLTDEQVQELCYRILHELRRGLAKDTHPKANVKCFVT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>HK</td>\n",
       "      <td>hexokinase</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>AACACTACCAATAAGTTGGAGTCATTACCCGCGGAATGTAAAATTT...</td>\n",
       "      <td>circular prokaryotic DNA</td>\n",
       "      <td>circular prokaryotic DNA</td>\n",
       "      <td>410144</td>\n",
       "      <td>411052</td>\n",
       "      <td>+</td>\n",
       "      <td>XGSSHHHHHHGSTKYALVGDVGGTNARLALCDIASGEISQAKTYSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>28</td>\n",
       "      <td>HK</td>\n",
       "      <td>hexokinase</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CCCTTCTAACCTCACCAGGCTGTTATCAGGAGAAATGAAGCAATAA...</td>\n",
       "      <td>10</td>\n",
       "      <td>10q22.1</td>\n",
       "      <td>69269991</td>\n",
       "      <td>69401882</td>\n",
       "      <td>+</td>\n",
       "      <td>MIAAQLLAYYFTELKDDQVKK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id name description          organism  \\\n",
       "0   4   HK  hexokinase        Drosophila   \n",
       "1  16   HK  hexokinase  Escherichia coli   \n",
       "2  28   HK  hexokinase      Homo sapiens   \n",
       "\n",
       "                                 nucleotide_sequence  \\\n",
       "0  ATGCCATTTGTGGACCCCTCAGCGTCGCACATATACACGCCATATC...   \n",
       "1  AACACTACCAATAAGTTGGAGTCATTACCCGCGGAATGTAAAATTT...   \n",
       "2  CCCTTCTAACCTCACCAGGCTGTTATCAGGAGAAATGAAGCAATAA...   \n",
       "\n",
       "                 chromosome                  location     start       end  \\\n",
       "0                         X                   1-28 cM   9585675   9589813   \n",
       "1  circular prokaryotic DNA  circular prokaryotic DNA    410144    411052   \n",
       "2                        10                   10q22.1  69269991  69401882   \n",
       "\n",
       "  strand                                translated_sequence  \n",
       "0      +  MVHELCQQLLLTDEQVQELCYRILHELRRGLAKDTHPKANVKCFVT...  \n",
       "1      +  XGSSHHHHHHGSTKYALVGDVGGTNARLALCDIASGEISQAKTYSG...  \n",
       "2      +                              MIAAQLLAYYFTELKDDQVKK  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"select genes by a specific enzyme: genes from different species for the HK enzyme\")\n",
    "df10 = pd.read_sql_query(\"select * from genes where id in (select gene_id from genes_in_enzymes where enzyme_id = 4)\", conn)\n",
    "df10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
